openedx/openedx-platform
# feat: Course mode admin explicitly defining upgrade deadline.

## Description

This PR adds a checkbox for `expiration_datetime_is_explicit` to the Django Admin's CourseMode edit page.  This allows Admin users to explicitly define an Upgrade Deadline date along with setting `expiration_datetime_is_explicit` to True in order to ensure that the Upgrade Deadline persists even if the course is published again.

**Before**
![Screen Shot 2022-11-02 at 13 34 42](https://user-images.githubusercontent.com/8311573/199561187-14539f07-2def-42c9-b6a3-46fbd799afe9.png)

**After**
![Screen Shot 2022-11-02 at 13 34 08](https://user-images.githubusercontent.com/8311573/199561189-91585689-c8f0-419b-b347-963fa6661150.png)

## Supporting information

This PR was the result of this initial investigation: https://discuss.openedx.org/t/course-mode-upgrade-deadline/8498.

## Testing instructions
I tested this using https://github.com/openedx/devstack which starts you out with some default courses already created.

1. Open Studio and navigate to your existing course.  Ensure that the course has an end date.
2. Using an existing course, go to `/admin/course_modes/coursemode` and click the ID of the associated record in order to navigate to the course mode edit page.
3. Edit the upgrade deadline date to be a date which is more than 10 days prior to the course end date from step 1.
4. Ensure that the checkbox for "Upgrade Deadline Explicitly Defined" is left unchecked/False.  Save the form.
5. Go back to Studio and update the start time (I updated the time by a single minute).  Save your changes.
6. Refresh the `/admin/course_modes/coursemode` page and you should now see that the upgrade deadline date for your course is 10 days prior to the course end date.
7. Repeat step 3 but this time checking/True the "Upgrade Deadline Explicitly Defined" checkbox.
8. Repeat step 5.
9. Refresh the `/admin/course_modes/coursemode` page and should now see that the upgrade deadline date is the value from step 7.

## Deadline

None.
